File: python/infinicore/ops/hardshrink.py
import infinicore
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def hardshrink(
    input: Tensor,
    lambd: float = 0.5,
    *,
    out=None,
) -> Tensor:
    r"""Applies the hardshrink function."""

    if infinicore.use_ntops and input.device.type in ("cuda", "musa"):
        return infinicore.ntops.torch.hardshrink(input, lambd)

    if out is None:
        return Tensor(_infinicore.hardshrink(input._underlying, lambd))
    _infinicore.hardshrink_(out._underlying, input._underlying, lambd)
    return out

File: src/infinicore/ops/hardshrink/hardshrink.cc
#include "infinicore/ops/hardshrink.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

common::OpDispatcher<Hardshrink::schema> &Hardshrink::dispatcher() {
    static common::OpDispatcher<Hardshrink::schema> dispatcher_;
    return dispatcher_;
}

void Hardshrink::execute(Tensor output, Tensor input, float lambd) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(output, input);
    infinicore::context::setDevice(input->device());
    dispatcher().lookup(input->device().getType())(output, input, lambd);
}

Tensor hardshrink(Tensor input, float lambd) {
    auto output = Tensor::empty(input->shape(), input->dtype(), input->device());
    hardshrink_(output, input, lambd);
    return output;
}

void hardshrink_(Tensor output, Tensor input, float lambd) {
    Hardshrink::execute(output, input, lambd);
}

} // namespace infinicore::op
